### PR TITLE
Transient version of ReentrancyGuard

### LINKED ISSUE #4205: `ReentrancyGuard` with transient opcodes `TSTORE` and `TLOAD`
**🧐 Motivation**

Since [EIP-1153](https://eips.ethereum.org/EIPS/eip-1153) is confirmed to be included in the upcoming Cancun-Deneb upgrade, I would like to discuss the possibility of using the future transient opcodes `TSTORE` and `TLOAD` for the `ReentrancyGuard` (abstract) contract.

Furthermore, I would like to highlight the initiatives (partially driven by myself), to disable reentrancy by default at the compiler level:
- **Vyper:** https://github.com/vyperlang/vyper/issues/3380
- **Solidity:** https://github.com/ethereum/solidity/issues/12996

### LINKED ISSUE #4888: Use transient storage in ReentrancyGuard
For backward compatibility, and support of chains that do no implement EIP-1153, we whould have two version (the current one + a new transient one). Current one would be marked as deprecated

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #4205
Fixes #4888
Replace #4955

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: contracts/utils/ReentrancyGuard.sol [hunk 1]
```diff
@@ -15,6 +15,9 @@ pragma solidity ^0.8.20;
  * those functions `private`, and then adding `external` `nonReentrant` entry
  * points to them.
  *
+ * NOTE: If EIP-1153 (transient storage) is available on the targeted network, you
+ * should consider using {TransientReentrancyGuard} instead.
```
Comment (ernestognw) [inline contracts/utils/ReentrancyGuard.sol:19]: I think this is more of a tip, also changed the redaction a bit.

```suggestion
 * TIP: If EIP-1153 (transient storage) is available on the chain you're deploying at,
 * consider using {ReentrancyGuardTransient} instead.
```

File: contracts/utils/README.adoc [hunk 2]
```diff
@@ -13,6 +13,7 @@ Miscellaneous contracts and libraries containing utility functions you can use t
  * {MerkleProof}: Functions for verifying https://en.wikipedia.org/wiki/Merkle_tree[Merkle Tree] proofs.
  * {EIP712}: Contract with functions to allow processing signed typed structure data according to https://eips.ethereum.org/EIPS/eip-712[EIP-712].
  * {ReentrancyGuard}: A modifier that can prevent reentrancy during certain functions.
+ * {ReentrancyGuardTransient}: Variant of {ReentrancyGuard} that uses transient storage (EIP-1153).
```
Comment (ernestognw) [inline contracts/utils/README.adoc:16]: I guess it makes sense to reference the EIP here.

```suggestion
 * {ReentrancyGuardTransient}: Variant of {ReentrancyGuard} that uses transient storage (https://eips.ethereum.org/EIPS/eip-1153[EIP-1153]).
```

**Review by ernestognw — APPROVED** [review]

**Comment by fuelfoundry** [comment]
Apologies if this has already been brought up elsewhere, but is anyone else getting:

Warning: Unreachable code.
  --> @openzeppelin/contracts/utils/ReentrancyGuard.sol:58:9:
   |
58 |         _nonReentrantAfter();
   |         ^^^^^^^^^^^^^^^^^^^^
   
Weirdly I only started seeing this today, it's balking at the modifier; gotta be the IDE just trippin. 

Respectfully,
-T

**Comment by Amxx** [comment]
@fuelfoundry there has been no change to the non-transient reentrancy guard in this PR.

Could you make a minimal reproductible example and open an issue ?

**Comment by fuelfoundry** [comment]
Fascinating, made a stripped down version of my current contract and everything is fine.  Will attempt to find what is causing it to bug out; if it's a function, variable, etc... and open an issue if it's reproducible and not some oversight on my part 😅.  Thank you. 🙏

**Pull request merged**